ProcessList.c

```diff
@@ -310,10 +310,29 @@ static void ProcessList_buildTree(ProcessList* this) {
       }
    }
 
+   // Under some ptrace(2) implementations, a process ptrace(2)-attaching a
+   // parent process in its process tree will causing the traced process to
+   // be re-parented to the tracing process; this will creating a loop in the
+   // process tree. In this case build separated tree(s) for loop(s) left
+   // here.
+   int display_size = Vector_size(this->displayList);
+   if(display_size < vsize) {
+      for (int i = 0; i < vsize; i++) {
+         Process* process = (Process*)Vector_get(this->processes, i);
+         if(!process->isRoot && Vector_indexOf(this->displayList, process, Process_pidEqualCompare) == -1) {
```

Comment:
Note that `Vector_indexOf` takes linear time, so this special case bumps tree build complexity to `O(n^2)`.

I'd suggest marking all processes as unseen in the root finding phase before sort by setting `tree_depth` to a guard value like `UINT_MAX` and checking processes for that guard value instead.

Reply:
You may also want to chase to loop in case the process you encounter first is not one of the loop processes, but a descendant of one of them.